### PR TITLE
39046 va-date Web Component

### PR DESCRIPTION
**Pull request by pdavies88**
## Chromatic

https://39046-date-web-component--60f9b557105290003b387cd5.chromatic.com

## Description
Closes https://github.com/department-of-veterans-affairs/va.gov-team/issues/39046

## Notes
This should not be merged in prior to the completion of `va-number-input`

## Testing done
Screen Shot 2022-04-13 at 9 46 26 AM

## Screenshots
Screen Shot 2022-04-13 at 9 46 37 AM

## Acceptance criteria
- [X] `va-date` is written and added to Storybook
- [X] E2E tests are written
- [ ] Add Chromatic link to #39047, assign ticket to Courtney, and move ticket to Current Sprint column
- [ ] Once accessibility review is complete, alert Carol that component is ready for staging review
- [ ] Address any staging review comments

## Definition of done
- [ ] Documentation has been updated, if applicable
- [X] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

### CODE REVIEW

**Comment by k80bowman** [comment]
This looks good to me. The only comment I have is that, when using voiceover on the Mac, it reads the first option for the month and day as "new line". It does this for the React component as well, so I'm not going to hold up the PR for this, but it might be nice to figure out a better option. I tried to do some research on how other design system date components handle this, but it seems that most other date components either only use text input boxes or are date pickers with a calendar attached.

File: packages/storybook/stories/va-select.stories.jsx [hunk 1]
```diff
@@ -77,7 +75,6 @@ const Template = ({
         value={value}
         required={required}
         error={error}
-        aria-live-region-text={ariaLiveRegionText}
```
Comment (bkjohnson) [inline packages/storybook/stories/va-select.stories.jsx:80]: Is this related to date things?

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+export class VaDate {
+  /**
+   * Render marker indicating field is required.
+   */
+  @Prop() required: boolean;
+
+  /**
+   * Label for the field.
+   */
+  @Prop() label!: string;
```
Comment (bkjohnson) [inline packages/web-components/src/components/va-date/va-date.tsx:29]: I like the `!` option - hopefully it will make storybook more useful.

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+  /**
+   * Set the default day on the day input.
+   */
+  @Prop() day: string;
+
+  /**
+   * Set the default year on the year input.
+   */
+  @Prop() year: string;
```
Comment (bkjohnson) [inline packages/web-components/src/components/va-date/va-date.tsx:65]: Can we also get a `value` prop since that's an established pattern for inputs? That would also let us use the familiar `event.target.value` pattern.

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+  yearChangeEvent: EventEmitter;
+
+  /**
+   * Fires when the year input loses focus
+   */
+  @Event({
+    composed: true,
+    bubbles: true,
+  })
+  yearBlurEvent: EventEmitter;
```
Comment (bkjohnson) [inline packages/web-components/src/components/va-date/va-date.tsx:119]: Are these custom events necessary? Also can we get a `dateChange` event or something? The current `Date` react component has the single event handler:

https://github.com/department-of-veterans-affairs/component-library/blob/8f70616a228f1a1a6d2f0c6980c0c1ac402e1c35/packages/react-components/src/components/Date/Date.jsx#L54

**Review by bkjohnson — COMMENTED** [review]
Some questions about the events.

File: packages/storybook/stories/va-select.stories.jsx (hunk 1 above)
Comment (pdavies88) [reply packages/storybook/stories/va-select.stories.jsx:80]: It was not in use and seems like we forgot to remove it.

Comment (pdavies88) [reply packages/storybook/stories/va-select.stories.jsx:80]: I will undo the change and write up a ticket for it instead

File: packages/react-components/src/helpers/options-for-select.js [hunk 5]
```diff
@@ -1,43 +1,3 @@
-export const maritalStatuses = [
```
Comment (pdavies88) [inline packages/react-components/src/helpers/options-for-select.js:1]: All of these are unused in Component Library and were copy and pasted from vets-website. Removing since it is unused code.

**Comment by pdavies88** [comment]
@bkjohnson @k80bowman @andresriveratoro this component has been reworked to only use a singular blur and change event for the component. It also takes in a singular prop for the date named value and value must be formatted as YYYY-MM-DD to behave correctly. 

Please take a look again and let me know what other steps would need to be taken to complete the development work for this component prior to a staging review.

**Comment by k80bowman** [comment]
The updates look great, thank you @pdavies88. I do have one unrelated question, is the year input supposed to have these up and down buttons? I don't see them in the React component:
image

**Comment by pdavies88** [comment]
@k80bowman they are there by default but CSS can be used to overwrite their appearance: https://www.w3schools.com/howto/howto_css_hide_arrow_number.asp
In the React Component they are being hidden via this code:
```
.usa-date-of-birth [type=number] {
    -moz-appearance: textfield;
}
```
I will update the CSS of the `va-number-input` to include that code so they no longer appear.

**Review by k80bowman — APPROVED** [review]
This looks good to me, thanks for all your work on this. And especially for adding the documentation around validation, that's great.

**Review by bkjohnson — COMMENTED** [review]
Mostly some comments about the interface for other engineers.

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+      maxYear,
+      minYear,
+      handleDateBlurEvent,
+      handleDateChangeEvent,
+      value,
+    } = this;
+
+    let { month, day, year } = this;
+
+    const [defaultYear, defaultMonth, defaultDay] = (value || '').split('-');
```
[the remark this answers is not in the document]
Comment (pdavies88) [reply packages/web-components/src/components/va-date/va-date.tsx]: I was running into console errors when trying to set a default value so I decided to leave as is

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+    const daysForSelectedMonth = month
+      ? days[month]
+      : parseInt(defaultMonth, 10) > 0
+      ? days[parseInt(defaultMonth, 10)]
+      : [];
```
[the remark this answers is not in the document]
Comment (pdavies88) [reply packages/web-components/src/components/va-date/va-date.tsx]: Comment has been added and code was condensed

**Comment by pdavies88** [comment]
@k80bowman @andresriveratoro @bkjohnson I have updated and reworked this PR again based on Brooks most recent comments. Please re-review the PR and let me know if additional changes are needed

File: packages/storybook/stories/va-date.stories.jsx [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+      <VaDate
+        label={label}
+        name={name}
+        required={required}
+        error={error}
+        value={value}
+        min-year={minYear}
+        max-year={maxYear}
+        onDateBlurEvent={e => console.log(e, 'DATE BLUR EVENT')}
+        onDateChangeEvent={e => setMonthDayYear(e.detail.path[0])}
```
[the remark this answers is not in the document]
Comment (pdavies88) [reply packages/storybook/stories/va-date.stories.jsx]: Code has been reworked to allow for just e.target.value to be used

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+  /**
+   * Set the `max` value on the year input.
+   */
+  @Prop() maxYear: number;
+
+  /**
+   * Set the default date value must be in YYYY-MM-DD format.
+   */
+  @Prop() value: string;
```
[the remark this answers is not in the document]
Comment (pdavies88) [reply packages/web-components/src/components/va-date/va-date.tsx]: Fixed

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+
+    const daysForSelectedMonth = month
+      ? days[month]
+      : parseInt(defaultMonth, 10) > 0
+      ? days[parseInt(defaultMonth, 10)]
+      : [];
+
+    let newDate = `${year || defaultYear}-${
+      (month && this.minTwoDigits(month)) || defaultMonth
+    }-${(day && this.minTwoDigits(day)) || defaultDay}`;
```
[the remark this answers is not in the document]
Comment (pdavies88) [reply packages/web-components/src/components/va-date/va-date.tsx]: Is have updated to use this thanks

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+    if (target.classList.contains('select-month')) {
+      this.month = target.value;
+    }
+    if (target.classList.contains('select-day')) {
+      this.day = target.value;
+    }
+    if (target.classList.contains('input-year')) {
+      this.year = target.value;
+    }
+    this.dateChangeEvent.emit(event);
```
[the remark this answers is not in the document]
Comment (pdavies88) [reply packages/web-components/src/components/va-date/va-date.tsx]: I have left a comment in the code and left the event firing as it currently does. Mainly it starts us on the path of controlling what is a valid date which might differ from team to team and I would like the app teams to control the validation aspect instead of our team.

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+  dateChangeEvent: EventEmitter;
+
+  /**
+   * Fires when the date input loses focus
+   */
+  @Event({
+    composed: true,
+    bubbles: true,
+  })
+  dateBlurEvent: EventEmitter;
```
[the remark this answers is not in the document]
Comment (pdavies88) [reply packages/web-components/src/components/va-date/va-date.tsx]: I have updated to remove the word event

File: packages/storybook/stories/va-date.stories.jsx [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+    if (target.classList.contains('select-month')) {
+      setMonthVal(target.value);
+    }
+    if (target.classList.contains('select-day')) {
+      setDayVal(target.value);
+    }
+    if (target.tagName === 'INPUT') {
+      setYearVal(target.value);
+    }
+  };
```
[the remark this answers is not in the document]
Comment (pdavies88) [reply packages/storybook/stories/va-date.stories.jsx]: Code has been reworked to allow for just e.target.value to be used

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 14]
```diff
@@ -5,6 +5,7 @@ import {
   days,
 } from '../../../../react-components/src/helpers/options-for-select.js';
 
+import { isFullDate } from '../../../../react-components/src/helpers/validations.js';
```
Comment (k80bowman) [inline packages/web-components/src/components/va-date/va-date.tsx:8]: We probably shouldn't be importing things like this from the react components. That could cause problems when we go to delete the React component, if whoever is doing that work doesn't know these imports are here. Could we copy these imports to a `date-utils.ts` file in the web components `utils` directory? Sorry I overlooked the months and days imports in my initial review.

Comment (pdavies88) [reply packages/web-components/src/components/va-date/va-date.tsx]: @k80bowman updates have been made and the `date-utils.ts` file has been created

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+
+    // Convert string to number to remove leading 0 when matching to days object
+    const daysForSelectedMonth =
+      parseInt(month, 10) > 0 ? days[parseInt(month, 10)] : [];
+
+    // Check validity of date if invalid provide message and error state styling
+    const dateInvalid =
+      required && (!isFullDate(value) || day > daysForSelectedMonth.length)
+        ? 'Please provide a valid date'
+        : null;
```
Comment (k80bowman) [inline packages/web-components/src/components/va-date/va-date.tsx:134]: Thank you for adding this in. Could you please add a test for it?

File: packages/web-components/src/utils/date-utils.ts [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+  8: thirtyOneDays,
+  9: thirtyDays,
+  10: thirtyOneDays,
+  11: thirtyDays,
+  12: thirtyOneDays,
+};
+
+export function isFullDate(date) {
+  return /\d{4}-\d{1,2}-\d{1,2}/.test(date);
+}
```
Comment (k80bowman) [inline packages/web-components/src/utils/date-utils.ts:130]: Thank you for copying all of this over. I know there isn't a test for this in the React components directory, but I'm wondering if maybe there should be here.

Comment (pdavies88) [reply packages/web-components/src/utils/date-utils.ts:130]: @k80bowman yes testing suite has been updated

File: packages/web-components/src/components/va-date/va-date.tsx (hunk 15 above)
Comment (pdavies88) [reply packages/web-components/src/components/va-date/va-date.tsx:134]: @k80bowman yes testing suite has been updated

File: packages/web-components/src/components/va-date/test/va-date.e2e.ts [hunk 17]
```diff
@@ -0,0 +1,147 @@
+import { newE2EPage } from '@stencil/core/testing';
+import { axeCheck } from '../../../testing/test-helpers';
+
+describe('va-date', () => {
```
Comment (bkjohnson) [inline packages/web-components/src/components/va-date/test/va-date.e2e.ts:4]: Can we also get a test showing that the event target's `value` has the updated date for the `dateChange` event?

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+      name,
+      error,
+      maxYear,
+      minYear,
+      handleDateBlur,
+      handleDateChange,
+      value,
+    } = this;
+
+    const [year, month, day] = (value || '').split('-');
```
Comment (bkjohnson) [inline packages/web-components/src/components/va-date/va-date.tsx:124]: After the split we could probably do a `.map()` which parses each value as an int.

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+
+    // Convert string to number to remove leading 0 when matching to days object
+    const daysForSelectedMonth =
+      parseInt(month, 10) > 0 ? days[parseInt(month, 10)] : [];
```
Comment (bkjohnson) [inline packages/web-components/src/components/va-date/va-date.tsx:128]: And that would simplify this.

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+            <span class="error-message" role="alert">
+              <span class="sr-only">Error</span> {error || dateInvalid}
+            </span>
+          )}
+          <div class="date-container">
+            <va-select
+              label="Month"
+              name={`${name}Month`}
+              // Remove starting 0 on days less than 10
+              value={month ? month.replace(/^0+/, '') : ''}
```
Comment (bkjohnson) [inline packages/web-components/src/components/va-date/va-date.tsx:158]: And maybe this as well

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+              label="Day"
+              name={`${name}Day`}
+              // If day value set is greater than amount of days in the month
+              // set to empty string instead
+              // Remove starting 0 on days less than 10
+              value={
+                daysForSelectedMonth.length < day
+                  ? ''
+                  : day
+                  ? day.replace(/^0+/, '')
```
Comment (bkjohnson) [inline packages/web-components/src/components/va-date/va-date.tsx:179]: Here too.

**Review by bkjohnson — COMMENTED** [review]
I think it would be good to have some tests for how the `dateChange` event works.

File: packages/web-components/src/components/va-date/test/va-date.e2e.ts (hunk 17 above)
Comment (pdavies88) [reply packages/web-components/src/components/va-date/test/va-date.e2e.ts:4]: I added to the test where I check for updates: https://github.com/department-of-veterans-affairs/component-library/pull/347/commits/601be1531e3ee837c5397954467269da578a0613#diff-999e0d6b4f145596771d22dcd3d694a5363c5f85675d8a905c4987e89c1cf631

File: packages/web-components/src/components/va-date/va-date.tsx (hunk 21 above)
Comment (pdavies88) [reply packages/web-components/src/components/va-date/va-date.tsx]: Change has been completed

File: packages/web-components/src/components/va-date/va-date.tsx (hunk 20 above)
Comment (pdavies88) [reply packages/web-components/src/components/va-date/va-date.tsx]: Change has been completed

File: packages/web-components/src/components/va-date/va-date.tsx (hunk 19 above)
Comment (pdavies88) [reply packages/web-components/src/components/va-date/va-date.tsx]: Change has been completed

File: packages/web-components/src/components/va-date/va-date.tsx (hunk 18 above)
Comment (pdavies88) [reply packages/web-components/src/components/va-date/va-date.tsx]: Change has been completed

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 22]
```diff
@@ -154,8 +155,8 @@ export class VaDate {
             <va-select
               label="Month"
               name={`${name}Month`}
-              // Remove starting 0 on days less than 10
-              value={month ? month.replace(/^0+/, '') : ''}
+              // Value must be a string
+              value={month ? month.toString() : ''}
```
Comment (bkjohnson) [inline packages/web-components/src/components/va-date/va-date.tsx:159]: Would `month?.toString()` work instead of having the ternary?

File: packages/web-components/src/components/va-date/test/va-date.e2e.ts (hunk 17 above)
Comment (bkjohnson) [reply packages/web-components/src/components/va-date/test/va-date.e2e.ts:4]: I still don't think we have a test for `dateChange` - I think it would be good to have one to test that it's firing when any 1 of the three components has a change.

File: packages/web-components/src/components/va-date/va-date.tsx (hunk 22 above)
Comment (pdavies88) [reply packages/web-components/src/components/va-date/va-date.tsx]: Change has been completed

**Comment by pdavies88** [comment]
@bkjohnson e2e tests to for the dateBlur and dateChange events have been added to the PR

**Review by bkjohnson — COMMENTED** [review]
Adding a `required` part to the select and number input so that we can set `display:none` when used inside of `<va-date>` means that others can do the same thing which I'm not too keen on. Is there a reason why we are doing that?

File: packages/web-components/src/components/va-number-input/va-number-input.tsx [hunk 23]
```diff
@@ -112,6 +116,8 @@ export class VaNumberInput {
           value={this.value}
           onInput={this.handleInput}
           onBlur={this.handleBlur}
+          part="input"
+          required={this.required || null}
```
Comment (bkjohnson) [inline packages/web-components/src/components/va-number-input/va-number-input.tsx:120]: Is this meant to be left in?

File: packages/web-components/src/components/va-select/va-select.tsx [hunk 24]
```diff
@@ -136,6 +136,8 @@ export class VaSelect {
           name={name}
           onKeyDown={() => this.handleKeyDown()}
           onChange={e => this.handleChange(e)}
+          part="select"
+          required={required || null}
```
Comment (bkjohnson) [inline packages/web-components/src/components/va-select/va-select.tsx:140]: Here as well.

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 25]
```diff
@@ -0,0 +1,206 @@
...
+      required && (!isFullDate(value) || day > daysForSelectedMonth.length)
+        ? 'Please provide a valid date'
+        : null;
+
+    // Setting new attribute to avoid conflicts with only using error attribute
+    // Error attribute should be leveraged for custom error messaging
+    // Fieldset has an implicit aria role of group
+    return (
+      <Host value={value} invalid={dateInvalid}>
+        <fieldset aria-label="Select Month and two digit day XX and four digit year format XXX">
```
Comment (bkjohnson) [inline packages/web-components/src/components/va-date/va-date.tsx:142]: @k80bowman we'll need to make sure we get a translation for this as well since this is new.

Comment (k80bowman) [reply packages/web-components/src/components/va-date/va-date.tsx:142]: Good call, I'll add it to the spreadsheet and alert the check-in team. Thanks!

Comment (bkjohnson) [reply packages/web-components/src/components/va-date/va-date.tsx:142]: We're also [missing an X](https://github.com/department-of-veterans-affairs/va.gov-team/issues/39047#issuecomment-1104468838).

```suggestion
        <fieldset aria-label="Select Month and two digit day XX and four digit year format XXXX">
```

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 26]
```diff
@@ -0,0 +1,206 @@
...
+          <legend>
+            {label || 'Date of birth'}{' '}
```
Comment (bkjohnson) [inline packages/web-components/src/components/va-date/va-date.tsx:144]: I know that the `Date of birth` thing is what's happening in the React component, but I don't know that there's an actual reason for it. It's just [been that way for over 3 years](https://github.com/department-of-veterans-affairs/component-library/blame/40e0142e0d28eb9a3d3556ffaac652595629babc/src/components/ErrorableDate/ErrorableDate.jsx#L89). Since the uses of `<Date>` and `<SimpleDate>` in `vets-website` are all passing label props, I think it might be good if we remove this rather than make an assumption that this component is for collecting date of birth.

File: packages/web-components/src/components/va-date/va-date.tsx [hunk 27]
```diff
@@ -0,0 +1,206 @@
...
+    const dateInvalid =
+      required && (!isFullDate(value) || day > daysForSelectedMonth.length)
+        ? 'Please provide a valid date'
+        : null;
+
+    // Setting new attribute to avoid conflicts with only using error attribute
+    // Error attribute should be leveraged for custom error messaging
+    // Fieldset has an implicit aria role of group
+    return (
+      <Host value={value} invalid={dateInvalid}>
```
Comment (bkjohnson) [inline packages/web-components/src/components/va-date/va-date.tsx:141]: Can you talk a bit more about the possible conflicts with `error` and `invalid`? I'm not sure why we have both.
